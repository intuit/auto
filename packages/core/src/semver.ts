import { ReleaseCalculationOptions } from "./types";
import * as t from "io-ts";

export type VersionLabel =
  | SEMVER.major
  | SEMVER.minor
  | SEMVER.patch
  | "skip"
  | "release";

enum SEMVER {
  major = "major",
  premajor = "premajor",
  minor = "minor",
  preminor = "preminor",
  patch = "patch",
  prepatch = "prepatch",
  noVersion = "",
}

export const preVersionMap = new Map([
  [SEMVER.major, SEMVER.premajor],
  [SEMVER.minor, SEMVER.preminor],
  [SEMVER.patch, SEMVER.prepatch],
]);

export type IVersionLabels = Map<VersionLabel | "none", string[]>;

export default SEMVER;

export const releaseLabels: VersionLabel[] = [
  SEMVER.major,
  SEMVER.minor,
  SEMVER.patch,
  "skip",
  "release",
];

/** Determine if a label is a label used for versioning */
export const isVersionLabel = (label: string): label is VersionLabel =>
  releaseLabels.includes(label as VersionLabel);

const labelDefinitionRequired = t.type({
  /** The label text */
  name: t.string,
});

const releaseType = t.union([
  t.literal("none"),
  t.literal("skip"),
  ...releaseLabels.map((l) => t.literal(l)),
]);
type ReleaseType = t.TypeOf<typeof releaseType>;

const labelDefinitionOptional = t.partial({
  /** A title to put in the changelog for the label */
  changelogTitle: t.string,
  /** The color of the label */
  color: t.string,
  /** The description of the label */
  description: t.string,
  /** What type of release this label signifies */
  releaseType,
  /** Whether to overwrite the base label */
  overwrite: t.boolean,
  /** Marks this label as the default label for unlabelled PRs */
  default: t.boolean,
});

export const labelDefinition = t.intersection([
  labelDefinitionOptional,
  labelDefinitionRequired,
]);
export type ILabelDefinition = t.TypeOf<typeof labelDefinition>;

const patchLabel: ILabelDefinition = {
  name: "patch",
  changelogTitle: "🐛 Bug Fix",
  description: "Increment the patch version when merged",
  releaseType: SEMVER.patch,
  color: "#870048",
};

export const defaultLabels: ILabelDefinition[] = [
  {
    name: "major",
    changelogTitle: "💥 Breaking Change",
    description: "Increment the major version when merged",
    releaseType: SEMVER.major,
    color: "#C5000B",
  },
  {
    name: "minor",
    changelogTitle: "🚀 Enhancement",
    description: "Increment the minor version when merged",
    releaseType: SEMVER.minor,
    color: "#F1A60E",
  },
  patchLabel,
  {
    name: "skip-release",
    description: "Preserve the current version when merged",
    releaseType: "skip",
    color: "#bf5416",
  },
  {
    name: "release",
    description: "Create a release when this pr is merged",
    releaseType: "release",
    color: "#007f70",
  },
  {
    name: "internal",
    changelogTitle: "🏠 Internal",
    description: "Changes only affect the internal API",
    releaseType: "none",
    color: "#696969",
  },
  {
    name: "documentation",
    changelogTitle: "📝 Documentation",
    description: "Changes only affect the documentation",
    releaseType: "none",
    color: "#cfd3d7",
  },
  {
    name: "tests",
    changelogTitle: "🧪 Tests",
    description: "Add or improve existing tests",
    releaseType: "none",
    color: "#ffd3cc",
  },
  {
    name: "dependencies",
    changelogTitle: "🔩 Dependency Updates",
    description: "Update one or more dependencies version",
    releaseType: "none",
    color: "#8732bc",
  },
  {
    name: "performance",
    changelogTitle: "🏎 Performance",
    description: "Improve performance of an existing feature",
    releaseType: SEMVER.patch,
    color: "#f4b2d8",
  },
];

/** Given two labels determine the next SEMVER bump. */
export function getHigherSemverTag(left: SEMVER, right: SEMVER): SEMVER {
  if (left === SEMVER.major || right === SEMVER.major) {
    return SEMVER.major;
  }

  if (left === SEMVER.minor || right === SEMVER.minor) {
    return SEMVER.minor;
  }

  if (left === SEMVER.patch || right === SEMVER.patch) {
    return SEMVER.patch;
  }

  return SEMVER.noVersion;
}

/** Get the semver bump for a release type */
export const getReleaseType = (releaseType?: ReleaseType) =>
  releaseType === "none" || releaseType === "skip"
    ? SEMVER.noVersion
    : releaseType === "release"
    ? SEMVER.patch
    : releaseType || SEMVER.patch;

/**
 * Determine the version bump from the labels on merged PRs.
 * Respects skip-release labels and the "onlyPublishWithReleaseLabel"
 * strategy.
 */
export function calculateSemVerBump(
  prLabels: string[][],
  labelMap: IVersionLabels,
  {
    onlyPublishWithReleaseLabel,
    labels = defaultLabels,
  }: ReleaseCalculationOptions & {
    /** The project's labels */
    labels?: ILabelDefinition[];
  } = {}
) {
  const defaultLabel = labels.find((l) => l.default) || patchLabel;
  const defaultReleaseType = defaultLabel.releaseType || SEMVER.patch;
  const releaseTypes = new Set<ReleaseType>();
  const skipReleaseLabels = labelMap.get("skip") || [];

  /** Find the release type + labels that match the given label */
  const getLabelEntry = (label: string) =>
    [...labelMap.entries()].find((pair) => pair[1].includes(label));

  prLabels.forEach((pr, index) => {
    // Default to a patch when:
    // 1. No labels on HEAD PR
    // 2. It has labels but none of them are auto labels
    if (
      index === 0 &&
      (pr.length === 0 || !pr.some((label) => Boolean(getLabelEntry(label))))
    ) {
      releaseTypes.add(defaultReleaseType);
    }

    pr.forEach((label) => {
      const userLabel = getLabelEntry(label);

      if (userLabel) {
        releaseTypes.add(userLabel[0]);
      }
    });
  });

  const lastMergedCommitLabels = prLabels[0] || [];
  console.log("Last Merged Commit Labels: ", lastMergedCommitLabels)
  const releaseLabels = labelMap.get("release") || [];
  console.log("Release Labels: ", releaseLabels)
  const skipRelease = onlyPublishWithReleaseLabel
    ? !lastMergedCommitLabels.some((label) => releaseLabels.includes(label))
    : lastMergedCommitLabels.some((label) => skipReleaseLabels.includes(label));
  console.log("skipRelease Value: ", skipRelease)
  if (skipRelease) {
    return SEMVER.noVersion;
  }

  // If PRs only have none or skip labels, skip the release
  const onlyNoReleaseLabels = [...releaseTypes].reduce(
    (condition, releaseType) =>
      condition && (releaseType === "none" || releaseType === "skip"),
    true
  );

  if (releaseTypes.size > 0 && onlyNoReleaseLabels) {
    return SEMVER.noVersion;
  }

  return [...releaseTypes]
    .map(getReleaseType)
    .reduce(getHigherSemverTag, getReleaseType(defaultReleaseType));
}
