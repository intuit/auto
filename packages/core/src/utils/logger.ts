import signale from "signale";

export type LogLevel = undefined | "verbose" | "veryVerbose" | "quiet";

export interface ILogger {
  /** The level at which to log messages */
  logLevel?: LogLevel;
  /** The default logger. Always on */
  log: signale.Signale<signale.DefaultMethods>;
  /** The verbose log. Has more debug logs */
  verbose: signale.Signale<signale.DefaultMethods>;
  /** The very verbose log. Has all debug logs */
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

/** Create a dummy logger for testing. */
export function dummyLog(): ILogger {
  return {
    log: new signale.Signale({ disabled: true }),
    verbose: new signale.Signale({ disabled: true }),
    veryVerbose: new signale.Signale({ disabled: true }),
  };
}

const loggerConfig: signale.SignaleOptions<"error" | "warn"> = { 
  types: { 
    error: { 
      stream: process.stderr,
      badge: '',
      label: '',
      color: ''
    },
    warn: {
      stream: process.stderr,
      badge: '',
      label: '',
      color: ''
    }
  }
}

const logger: ILogger = {
  log: new signale.Signale(loggerConfig),
  verbose: new signale.Signale(loggerConfig),
  veryVerbose: new signale.Signale(loggerConfig),
};

/** Turn the logs on an off */
function toggleLogs(
  options: Record<Exclude<keyof ILogger, "logLevel">, boolean>
) {
  Object.entries(options).forEach(([level, enabled]) => {
    if (enabled) {
      // @ts-ignore
      logger[level].enable();
    } else {
      // @ts-ignore
      logger[level].disable();
    }
  });
}

/** Set the log level */
export function setLogLevel(newLogLevel: LogLevel) {
  logger.logLevel = newLogLevel;

  if (logger.logLevel === "verbose") {
    toggleLogs({ log: true, verbose: true, veryVerbose: false });
  } else if (logger.logLevel === "veryVerbose") {
    toggleLogs({ log: true, verbose: true, veryVerbose: true });
  } else if (logger.logLevel === "quiet") {
    toggleLogs({ log: false, verbose: false, veryVerbose: false });
  } else {
    toggleLogs({ log: true, verbose: false, veryVerbose: false });
  }
}

/** Create a logger the the given log level. */
export default function createLog(): ILogger {
  return logger;
}
