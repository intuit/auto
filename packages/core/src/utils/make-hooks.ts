import {
  AsyncParallelHook,
  AsyncSeriesBailHook,
  AsyncSeriesWaterfallHook,
  SyncHook,
  AsyncSeriesHook,
} from "tapable";
import { IAutoHooks } from "../auto";
import { IChangelogHooks } from "../changelog";
import { ILogParseHooks } from "../log-parse";
import { IReleaseHooks } from "../release";
import { InteractiveInitHooks } from "../init";

/** Make the hooks for "auto" */
export const makeHooks = (): IAutoHooks => ({
  beforeRun: new AsyncSeriesHook(["config"]),
  afterRun: new AsyncSeriesHook(["config"]),
  modifyConfig: new AsyncSeriesWaterfallHook(["config"]),
  validateConfig: new AsyncSeriesBailHook(["name", "options"]),
  beforeShipIt: new AsyncSeriesHook(["context"]),
  afterChangelog: new AsyncSeriesHook(["context"]),
  beforeCommitChangelog: new AsyncSeriesHook(["context"]),
  afterShipIt: new AsyncParallelHook(["context"]),
  makeRelease: new AsyncSeriesBailHook(["releaseInfo"]),
  afterRelease: new AsyncParallelHook(["releaseInfo"]),
  onCreateRelease: new SyncHook(["options"]),
  onCreateChangelog: new SyncHook(["changelog", "version"]),
  onCreateLogParse: new SyncHook(["logParse"]),
  getAuthor: new AsyncSeriesBailHook(),
  getPreviousVersion: new AsyncSeriesBailHook(),
  getRepository: new AsyncSeriesBailHook(),
  prCheck: new AsyncSeriesBailHook(["prInformation"]),
  beforeVersion: new AsyncSeriesHook(["context"]),
  version: new AsyncSeriesHook(["version"]),
  afterVersion: new AsyncSeriesHook(["context"]),
  publish: new AsyncSeriesHook(["version"]),
  afterPublish: new AsyncSeriesHook(),
  canary: new AsyncSeriesBailHook(["canaryContext"]),
  next: new AsyncSeriesWaterfallHook(["preReleaseVersions", "context"]),
});

/** Make the hooks for "Release" */
export const makeReleaseHooks = (): IReleaseHooks => ({
  onCreateChangelog: new SyncHook(["changelog", "version"]),
  createChangelogTitle: new AsyncSeriesBailHook([]),
  onCreateLogParse: new SyncHook(["logParse"]),
});

/** Make the hooks for "LogParse" */
export const makeLogParseHooks = (): ILogParseHooks => ({
  parseCommit: new AsyncSeriesWaterfallHook(["commit"]),
  omitCommit: new AsyncSeriesBailHook(["commit"]),
});

/** Make the hooks for "InteractiveInit" */
export const makeInteractiveInitHooks = (): InteractiveInitHooks => ({
  writeRcFile: new AsyncSeriesBailHook(["rcFile"]),
  getRepo: new AsyncSeriesBailHook([]),
  getAuthor: new AsyncSeriesBailHook([]),
  createEnv: new AsyncSeriesWaterfallHook(["vars"]),
  configurePlugin: new AsyncSeriesBailHook(["pluginName"]),
});

/** Make the hooks for "Changelog" */
export const makeChangelogHooks = (): IChangelogHooks => ({
  addToBody: new AsyncSeriesWaterfallHook(["notes", "commits"]),
  sortChangelogLines: new AsyncSeriesWaterfallHook(["lines"]),
  renderChangelogLine: new AsyncSeriesWaterfallHook(["line", "commit"]),
  renderChangelogTitle: new AsyncSeriesBailHook(["commits", "lineRender"]),
  renderChangelogAuthor: new AsyncSeriesBailHook([
    "author",
    "commit",
    "options",
  ]),
  renderChangelogAuthorLine: new AsyncSeriesBailHook(["author", "user"]),
  omitReleaseNotes: new AsyncSeriesBailHook(["commit"]),
});
