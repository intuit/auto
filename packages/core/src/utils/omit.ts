import fromEntries from "fromentries";

/** Omit properties from an object */
export const omit = <Props, Prop extends keyof Props>(
  obj: Props,
  keys: Prop[]
) =>
  fromEntries(
    Object.entries(obj as {}).filter(([key]) => !keys.includes(key as Prop))
  ) as Omit<Props, Prop>;
