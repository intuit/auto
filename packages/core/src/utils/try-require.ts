import importCwd from "import-cwd";
import requireg from "requireg";
import importFrom from "import-from";
import createLog from "./logger";

const logger = createLog();

/** Try to require something either from the CWD or the regular way */
export default function tryRequire(tryPath: string, from?: string) {
  try {
    // Require from CWD
    return importCwd(tryPath);
  } catch (error) {
    // if we try to actually require npm we will import something that is the actual npm API
    // not the plugin that we want
    if (tryPath === "npm") {
      return;
    }

    // If a plugin has any errors we want to inform the user
    if (
      (!error && typeof error !== "object") ||
      (error &&
        typeof error === "object" &&
        "code" in error &&
        error.code !== "MODULE_NOT_FOUND")
    ) {
      throw error;
    }

    try {
      // Require from __dirname and then common global package places. Needed for npx and global installs
      return requireg(tryPath);
    } catch (error) {
      if (error && typeof error === "object" && "message" in error) {
        logger.veryVerbose.warn(error.message);
      }
    }

    if (from) {
      try {
        // For loading plugins specified in a configuration (yarn 2)
        return importFrom(from, tryPath);
      } catch (error) {
        logger.veryVerbose.warn({ from, tryPath, error });
      }
    }
  }
}
