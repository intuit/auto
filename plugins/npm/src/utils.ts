import * as fs from 'fs';
import path from 'path';
import { promisify } from 'util';

export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);

/** Load and parse the root package json for the project */
export async function loadPackageJson(root = './'): Promise<IPackageJSON> {
  return JSON.parse(await readFile(path.join(root, 'package.json'), 'utf-8'));
}
