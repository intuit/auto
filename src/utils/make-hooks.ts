import { AsyncSeriesBailHook, AsyncSeriesHook, SyncHook } from 'tapable';
import { IGitHubReleaseHooks } from '../github-release';
import { ILogParseHooks } from '../log-parse';
import { IAutoHooks } from '../main';

export const makeHooks = (): IAutoHooks => ({
  beforeRun: new SyncHook(['config']),
  beforeShipIt: new SyncHook([]),
  onCreateGitHubRelease: new SyncHook(['config']),
  onCreateLogParse: new SyncHook(['gitHubReleaseConfig']),
  getAuthor: new AsyncSeriesBailHook([]),
  getPreviousVersion: new AsyncSeriesBailHook(['prefixRelease']),
  getRepository: new AsyncSeriesBailHook([]),
  publish: new AsyncSeriesHook(['version'])
});

export const makeGitHubReleaseHooks = (): IGitHubReleaseHooks => ({
  onCreateLogParse: new SyncHook(['logParser'])
});

export const makeLogHooks = (): ILogParseHooks => ({
  renderChangelogLine: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogTitle: new AsyncSeriesBailHook(['commits', 'lineRender']),
  renderChangelogAuthor: new AsyncSeriesBailHook([
    'author',
    'commit',
    'options'
  ]),
  renderChangelogAuthorLine: new AsyncSeriesBailHook(['author', 'user'])
});
