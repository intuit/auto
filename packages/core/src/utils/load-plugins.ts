/* eslint-disable @typescript-eslint/no-explicit-any */

import * as path from 'path';
import Auto from '../auto';
import { ILogger } from './logger';
import tryRequire from './try-require';
import InteractiveInit from '../init';

export type IPluginConstructor = new (options?: any) => IPlugin;

/** A plugin to auto */
export interface IPlugin {
  /** The name to identify the plugin by */
  name: string;
  /** Called when running `auto init`. gives plugin ability to add custom init experience. */
  init?(initializer: InteractiveInit): void;
  /** Called when registering the plugin with auto */
  apply(auto: Auto): void;
}

/** Require a plugin and log where it was found. */
function requirePlugin(pluginPath: string, logger: ILogger) {
  const plugin = tryRequire(pluginPath);

  if (plugin) {
    logger.verbose.info(`Found plugin using: ${pluginPath}`);
  }

  return plugin as IPluginConstructor;
}

/** Try to load a plugin in various ways */
export default function loadPlugin(
  [pluginPath, options]: [string, any],
  logger: ILogger
): IPlugin | undefined {
  let plugin:
    | IPluginConstructor
    | {
        /** The plugin under the default export */
        default: IPluginConstructor;
      }
    | undefined;

  // Try requiring a path
  if (pluginPath.startsWith('.') || pluginPath.startsWith('/')) {
    plugin = requirePlugin(pluginPath, logger);
  }

  // Try requiring a path from cwd
  if (!plugin && (pluginPath.startsWith('.') || pluginPath.startsWith('/'))) {
    const localPath = path.join(process.cwd(), pluginPath);
    plugin = requirePlugin(localPath, logger);

    if (!plugin) {
      logger.log.warn(`Could not find plugin from path: ${localPath}`);
      return;
    }
  }

  // For pkg bundle
  if (!plugin) {
    const pkgPath = path.join(
      __dirname,
      '../../../../../plugins/',
      pluginPath,
      'dist/index.js'
    );
    plugin = requirePlugin(pkgPath, logger);
  }

  // For a user created plugin
  if (!plugin) {
    plugin = requirePlugin(`auto-plugin-${pluginPath}`, logger);
  }

  // Try importing official plugin
  if (!plugin) {
    plugin = requirePlugin(path.join('@auto-it', pluginPath), logger);
  }

  // Try requiring a package
  if (!plugin && pluginPath.includes('auto-plugin-')) {
    plugin = requirePlugin(pluginPath, logger);
  }

  if (!plugin) {
    logger.log.warn(`Could not find plugin: ${pluginPath}`);
    return;
  }

  try {
    if ('default' in plugin && plugin.default) {
      // eslint-disable-next-line new-cap
      return new plugin.default(options);
    }

    return new (plugin as IPluginConstructor)(options);
  } catch (error) {
    logger.log.error(
      `Plugin at the following path encountered an error: ${pluginPath}`
    );
    throw error;
  }
}
