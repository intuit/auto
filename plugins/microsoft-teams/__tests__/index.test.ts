import Auto from "@auto-it/core";
import makeCommitFromMsg from "@auto-it/core/src/__tests__/make-commit-from-msg";
import { dummyLog } from "@auto-it/core/src/utils/logger";
import { makeHooks } from "@auto-it/core/src/utils/make-hooks";
import { defaultLabels } from "@auto-it/core/dist/semver";
import { execSync } from "child_process";
import createHttpsProxyAgent from "https-proxy-agent";

import MicrosoftTeamsPlugin from "../src";

const fetchSpy = jest.fn();
// @ts-ignore
jest.mock("node-fetch", () => (...args) => {
  fetchSpy(...args);
});

beforeEach(() => {
  fetchSpy.mockClear();
});

const mockResponse = [
  {
    data: {
      html_url: "https://git.hub/some/project/releases/v1.0.0",
      name: "v1.0.0",
    },
  },
];

// For the purpose of this test, we use the current branch as the "prerelease" branch to fake being on a "next" branch
const nextBranch = execSync("git rev-parse --abbrev-ref HEAD", {
  encoding: "utf8",
}).trim();

const mockAuto = {
  git: {},
  logger: dummyLog(),
} as any;

describe("createPost", () => {
  test("doesn't post with no new version", async () => {
    const plugin = new MicrosoftTeamsPlugin("https://custom-microsoft-url");
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({ hooks } as Auto);

    await hooks.afterRelease.promise({
      lastRelease: "0.1.0",
      commits: [],
      releaseNotes: "# My Notes",
    });

    expect(plugin.createPost).not.toHaveBeenCalled();
  });

  test("doesn't post with no commits", async () => {
    const plugin = new MicrosoftTeamsPlugin("https://custom-microsoft-url");
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({ hooks, options: {} } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [],
      releaseNotes: "# My Notes",
    });

    expect(plugin.createPost).not.toHaveBeenCalled();
  });

  test("doesn't post with skip release label", async () => {
    const plugin = new MicrosoftTeamsPlugin("https://custom-microsoft-url");
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({
      hooks,
      options: {},
      config: { labels: defaultLabels },
    } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("skipped", { labels: ["skip-release"] })],
      releaseNotes: "# My Notes",
    });

    expect(plugin.createPost).not.toHaveBeenCalled();
  });

  test("doesn't post without url", async () => {
    // @ts-ignore
    const plugin = new MicrosoftTeamsPlugin({ url: undefined });
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({ hooks, options: {} } as Auto);

    await expect(
      hooks.afterRelease.promise({
        newVersion: "1.0.0",
        lastRelease: "0.1.0",
        commits: [makeCommitFromMsg("a patch")],
        releaseNotes: "# My Notes",
      })
    ).rejects.toBeInstanceOf(Error);
  });

  test("doesn't post when prerelease branch and using default prereleasePublish setting", async () => {
    // @ts-ignore
    const plugin = new MicrosoftTeamsPlugin({
      url: "https://custom-microsoft-url",
    });
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({
      ...mockAuto,
      hooks,
      options: {},
      config: {
        prereleaseBranches: [nextBranch],
        labels: defaultLabels,
      },
    } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("a patch")],
      releaseNotes: "# My Notes",
    });
    expect(fetchSpy).not.toHaveBeenCalled();
  });

  test("doesn't post when prerelease branch setting is false", async () => {
    // @ts-ignore
    const plugin = new MicrosoftTeamsPlugin({
      url: "https://custom-microsoft-url",
      publishPreRelease: false,
    });
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({
      ...mockAuto,
      hooks,
      options: {},
      config: {
        prereleaseBranches: [nextBranch],
        labels: defaultLabels,
      },
    } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("a patch")],
      releaseNotes: "# My Notes",
    });
    expect(fetchSpy).not.toHaveBeenCalled();
  });

  test("posts when prerelease branch setting is true", async () => {
    // @ts-ignore
    const plugin = new MicrosoftTeamsPlugin({
      url: "https://custom-microsoft-url",
      publishPreRelease: true,
    });
    const hooks = makeHooks();

    jest.spyOn(plugin, "createPost").mockImplementation();
    // @ts-ignore
    plugin.apply({
      ...mockAuto,
      hooks,
      options: {},
      config: { prereleaseBranches: ["next"], labels: defaultLabels },
    } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("a patch")],
      releaseNotes: "# My Notes",
      // @ts-ignore
      response: mockResponse,
    });
    expect(plugin.createPost).toHaveBeenCalledTimes(1);
  });

  test("should call Microsoft Office Teams api with minimal config", async () => {
    const plugin = new MicrosoftTeamsPlugin("https://custom-microsoft-url");

    await plugin.createPost(
      mockAuto,
      "# My Notes\n- PR [some link](google.com)",
      "*<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*",
      undefined
    );

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe(
      "https://custom-microsoft-url"
    );
    expect(fetchSpy.mock.calls[0][1].agent).toBeUndefined();
    expect(fetchSpy.mock.calls[0][1].body).toMatchSnapshot();
  });

  test("should call Microsoft Office Teams api through http proxy", async () => {
    const plugin = new MicrosoftTeamsPlugin("https://custom-microsoft-url");
    process.env.http_proxy = "http-proxy";

    await plugin.createPost(
      mockAuto,
      "# My Notes\n- PR [some link](google.com)",
      "*<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*",
      createHttpsProxyAgent("mock-url")
    );

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe(
      "https://custom-microsoft-url"
    );
    expect(fetchSpy.mock.calls[0][1].agent).not.toBeUndefined();
    expect(fetchSpy.mock.calls[0][1].body).toMatchSnapshot();
  });

  test("should call Microsoft Office Teams api through https proxy", async () => {
    const plugin = new MicrosoftTeamsPlugin("https://custom-microsoft-url");
    process.env.https_proxy = "https-proxy";

    await plugin.createPost(
      mockAuto,
      "# My Notes\n- PR [some link](google.com)",
      "*<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*",
      createHttpsProxyAgent("mock-url")
    );

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe(
      "https://custom-microsoft-url"
    );
    expect(fetchSpy.mock.calls[0][1].agent).not.toBeUndefined();
    expect(fetchSpy.mock.calls[0][1].body).toMatchSnapshot();
  });

  test("should call Microsoft Office Teams api", async () => {
    const plugin = new MicrosoftTeamsPlugin({ url: "https://custom-microsoft-url" });
    const hooks = makeHooks();
    plugin.apply({ hooks, options: {}, ...mockAuto } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("a patch")],
      releaseNotes: "# My Notes\n- PR [some link](google.com)",
      // @ts-ignore
      response: mockResponse,
    });

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe(
      "https://custom-microsoft-url"
    );
    expect(fetchSpy.mock.calls[0][1].body).toMatchSnapshot();
  });


  // test("should call Microsoft Office Teams api with custom atTarget", async () => {
  //   const plugin = new MicrosoftTeamsPlugin({
  //     url: "https://custom-microsoft-url",
  //     atTarget: "here",
  //   });
  //   const hooks = makeHooks();
  //   plugin.apply({ hooks, options: {}, ...mockAuto } as Auto);

  //   await hooks.afterRelease.promise({
  //     newVersion: "1.0.0",
  //     lastRelease: "0.1.0",
  //     commits: [makeCommitFromMsg("a patch")],
  //     releaseNotes: "# My Notes\n- PR [some link](google.com)",
  //     // @ts-ignore
  //     response: mockResponse,
  //   });

  //   expect(fetchSpy).toHaveBeenCalled();
  //   expect(fetchSpy.mock.calls[0][0]).toBe(
  //     "https://custom-microsoft-url"
  //   );
  //   expect(fetchSpy.mock.calls[0][1].body.includes("@here")).toBe(true);
  //   expect(fetchSpy.mock.calls[0][1].body).toMatchSnapshot();
  // });

  test("should call Microsoft Office Teams api in env var", async () => {
    process.env.MICROSOFT_TEAMS_WEBHOOK_URL = "https://foo.bar";
    const plugin = new MicrosoftTeamsPlugin();
    const hooks = makeHooks();
    plugin.apply({ hooks, options: {}, ...mockAuto } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("a patch")],
      releaseNotes: "# My Notes\n- PR [some link](google.com)",
      // @ts-ignore
      response: mockResponse,
    });

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][0]).toBe("https://foo.bar");
    expect(fetchSpy.mock.calls[0][1].body).toMatchSnapshot();
  });

  test("should add title", async () => {
    process.env.MICROSOFT_TEAMS_WEBHOOK_URL = "https://foo.bar";
    const plugin = new MicrosoftTeamsPlugin({ title: "My Cool Project" });
    const hooks = makeHooks();
    plugin.apply({ hooks, options: {}, ...mockAuto } as Auto);

    await hooks.afterRelease.promise({
      newVersion: "1.0.0",
      lastRelease: "0.1.0",
      commits: [makeCommitFromMsg("a patch")],
      releaseNotes: "# My Notes\n- PR [some link](google.com)",
      // @ts-ignore
      response: mockResponse,
    });

    expect(fetchSpy).toHaveBeenCalled();
    expect(fetchSpy.mock.calls[0][1].body).toMatchSnapshot();
  });
});
