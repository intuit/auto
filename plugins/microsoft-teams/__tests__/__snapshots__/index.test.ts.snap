// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPost should add title 1`] = `"{\\"@context\\":\\"http://schema.org/extensions\\",\\"@type\\":\\"MessageCard\\",\\"text\\":\\"# My Notes/n- PR [some link](google.com)\\",\\"potentialAction\\":[{\\"@type\\":\\"OpenUri\\",\\"name\\":\\"Learn More\\",\\"targets\\":[{\\"os\\":\\"default\\",\\"uri\\":\\"*<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\"}]}]}"`;

exports[`createPost should call Microsoft Office Teams api 1`] = `"{\\"@context\\":\\"http://schema.org/extensions\\",\\"@type\\":\\"MessageCard\\",\\"text\\":\\"# My Notes/n- PR [some link](google.com)\\",\\"potentialAction\\":[{\\"@type\\":\\"OpenUri\\",\\"name\\":\\"Learn More\\",\\"targets\\":[{\\"os\\":\\"default\\",\\"uri\\":\\"*<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\"}]}]}"`;

exports[`createPost should call Microsoft Office Teams api in env var 1`] = `"{\\"@context\\":\\"http://schema.org/extensions\\",\\"@type\\":\\"MessageCard\\",\\"text\\":\\"# My Notes/n- PR [some link](google.com)\\",\\"potentialAction\\":[{\\"@type\\":\\"OpenUri\\",\\"name\\":\\"Learn More\\",\\"targets\\":[{\\"os\\":\\"default\\",\\"uri\\":\\"*<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\"}]}]}"`;

exports[`createPost should call Microsoft Office Teams api through http proxy 1`] = `"{\\"@context\\":\\"http://schema.org/extensions\\",\\"@type\\":\\"MessageCard\\",\\"text\\":\\"# My Notes/n- PR [some link](google.com)\\",\\"potentialAction\\":[{\\"@type\\":\\"OpenUri\\",\\"name\\":\\"Learn More\\",\\"targets\\":[{\\"os\\":\\"default\\",\\"uri\\":\\"*<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\"}]}]}"`;

exports[`createPost should call Microsoft Office Teams api through https proxy 1`] = `"{\\"@context\\":\\"http://schema.org/extensions\\",\\"@type\\":\\"MessageCard\\",\\"text\\":\\"# My Notes/n- PR [some link](google.com)\\",\\"potentialAction\\":[{\\"@type\\":\\"OpenUri\\",\\"name\\":\\"Learn More\\",\\"targets\\":[{\\"os\\":\\"default\\",\\"uri\\":\\"*<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\"}]}]}"`;

exports[`createPost should call Microsoft Office Teams api with minimal config 1`] = `"{\\"@context\\":\\"http://schema.org/extensions\\",\\"@type\\":\\"MessageCard\\",\\"text\\":\\"# My Notes/n- PR [some link](google.com)\\",\\"potentialAction\\":[{\\"@type\\":\\"OpenUri\\",\\"name\\":\\"Learn More\\",\\"targets\\":[{\\"os\\":\\"default\\",\\"uri\\":\\"*<https://git.hub/some/project/releases/v1.0.0|v1.0.0>*\\"}]}]}"`;
